### DISCUSSION TITLE
Python Closures: A way for RNS callbacks to access outer scope

### DISCUSSION THREAD

**Discussion by ThomasFreedman** [markqvist/Reticulum#314] [General]
Let me start out by saying I avoid global variables like the "plague". 

The first time I had to figure out how to handle this was with the announce_handler. Since the RNS API required a _class_ and not a function reference, it was easy to add a "parent" parameter to the class constructor to pass data in or out of the announce handler's received_announce method via self.parent.

It wasn't so easy with various RNS callbacks however. I didn't see a way to reference the class context of a callback method until I learned about [Python closures](https://www.geeksforgeeks.org/python-closures/). All of my Reticulum code is within a class definition, so this technique is perfect for this case and it doesn't require any global variables.

Here's a snippet example for RNS. Using the closure concept, the linkCreated inner function has access to the outer function's "self" variable, and thus access to all of the outer class objects & properties.

```
    def createLinkToServer(self, svr):
        # These callbacks will be called when server links are opened or closed.
        def linkCreated(link) -> None:
            RNS.log(f"Connected to {link.destination}")
            link.identify(self.Svr_ID)  # Expose our identiity to server

        # Garbage collect the server from the dictionary when the link is closed
        def linkClosed(link) -> None:
            RNS.log(f"Disconnected from {link.destination}")
                for s in self.Servers.keys():
                    if self.Servers[s]["link"] == link:
                        del self.Servers[s]
                        
        # Create a link to the server
        prnt = svr["prnt"]
        if svr["path"]:   # A link requires a path; this only says if it exists
            RNS.log(f"createLinkToServer: Creating a link to {prnt}...")
            link = RNS.Link(svr["dest"], established_callback=linkCreated,
                                         closed_callback=linkClosed)
            svr["link"] = link
        else: self.checkPathForLink(svr) # Handle the "no path" case

```

**Comment by markqvist** [comment]
Yes, using closures, or anonymous functions as they are sometimes called, is a good way to do this!
